blocks/api/categories.js

```diff
@@ -20,10 +23,111 @@ const categories = [
 ];
 
 /**
- * Returns all the block categories
+ * Returns all the block categories.
  *
- * @return {Array} Block categories
+ * @return {Array} categories
  */
 export function getCategories() {
 	return categories;
 }
+
+/**
+ * Register a new block Category.
+ *
+ * @param {Array} category e.g {slug: 'custom', title: __('Custom Blocks')}
+ *
+ * @return {Array} categories
+ */
+export function registerCategory( category ) {
+	if ( ! category ) {
+		console.error(
+			'The block Category must be defined'
+		);
+		return;
+	}
+	if ( ! category.slug ) {
+		console.error(
+			'The block Category slug must be defined'
+		);
+		return;
+	}
+	if ( ! /^[a-z0-9-]+$/.test( category.slug ) ) {
+		console.error(
+			'The block Category slug must not contain characters which are invalid for urls'
+		);
+		return;
+	}
+	if ( categories.find( x => x.slug === category.slug ) ) {
+		console.error(
+			'The block Category "' + category.slug + '" is already registered'
+		);
+		return;
+	}
+	if ( ! category.title ) {
+		console.error(
+			'The block Category title must be defined'
+		);
+		return;
+	}
+
+	categories.push( category );
+	return categories;
+}
+
+/**
+ *
+ * Get sorted categories by property.
+ *
+ * @param {String} sortProperty The key to sort by
+ *
+ * @returns {Array} categories
+ */
+
+export function getSortedCategories( sortProperty ) {
+	if ( ! sortProperty ) {
+		console.error(
+			'The sortProperty must be defined'
+		);
+		return;
+	}
+	if ( typeof sortProperty !== 'string' ) {
+		console.error(
+			'The sortProperty must be a string'
+		);
+		return;
+	}
+	categories = sortBy( categories, sortProperty );
```

Comment:
Might want to verify here that this sort sorts as expected. JavaScript's default sort doesn't always produce the results we want, but when comparing strings we are thankfully given a function to make it more reliable…

```js
const byGivenProperty = ( a, b ) => a[ sortProperty ].localeCompare( b[ sortProperty ] );
categories.sort( byGivenProperty )
```

Reply:
Hi @dmsnell,
I'm using the loadash sortBy which already handles exceptions.
In this specific case it protects the sorting even if the order hasn't set for the other elements in the categories array. The built-in sort doesn't handle that exception.

Antonio